### ISSUE TITLE
Formal verification of the abstract cryptographic protocol

### ISSUE THREAD

**Issue by defuse** [zcash/zcash#5049]
Investigate the feasibility of formally verifying some properties of the abstract cryptographic protocol, i.e. just the Sapling/Orchard protocol and not necessarily the entire the proving system. Given the expertise that has designed/reviewed it already, this is probably the only way to find bugs, and it'll mean we can ship new features quicker and safer.

I can't imagine it would be *that* much harder than verifying the Signal protocol, which has been done in [Automated Verification for Secure Messaging Protocols and Their Implementations: A Symbolic and Computational Approach](https://prosecco.gforge.inria.fr/personal/bblanche/publications/KobeissiBhargavanBlanchetEuroSP17.pdf). See also @nadimkobeissi's [thesis on the topic](https://hal.inria.fr/tel-01950884/). I'm not sure what other research is relevant.

A security proof for Sapling can be found [here](https://github.com/zcash/sapling-security-analysis/blob/master/SaplingSecurityProof.pdf), suggesting it's at least within the realm of possibility.

**Comment by defuse** [comment]
https://prosecco.gforge.inria.fr/personal/bblanche/publications/BlanchetFnTPS16.pdf

**Comment by nadimkobeissi** [comment]
It would be a pleasure to help with this, if my contributions are welcome.

**Comment by defuse** [comment]
@nadimkobeissi yes, absolutely welcome! You might be able to get funding for it even, through [ZOMG Grants](https://zcashomg.org/). I'd strongly advocate in favor of funding a grant for this kind of work. Please let me know if there's anything I can do or help with to encourage this kind of thing :)

**Comment by nadimkobeissi** [comment]
@defuse Thank you!

I'd like to understand your goals clearly. Is the goal focused on formally modeling and verifying the Sapling protocol? Where can I access the most complete and up to date protocol specification?
